pallets/omnipool/src/lib.rs

```diff
@@ -1659,4 +1673,111 @@ impl<T: Config> Pallet<T> {
 
 		Err(Error::<T>::NotAllowed.into())
 	}
+
+	/// Remove asset from list of Omnipool assets.
+	pub fn remove_asset(asset_id: T::AssetId) -> DispatchResult {
+		<Assets<T>>::remove(asset_id);
+		Ok(())
+	}
+
+	/// Insert or update position with given position data.
+	pub fn set_position(position_id: T::PositionItemId, position: &Position<Balance, T::AssetId>) -> DispatchResult {
+		<Positions<T>>::insert(position_id, position);
+		Ok(())
+	}
+
+	/// Add new asset to list of Omnipool assets.
+	pub fn add_asset(asset_id: T::AssetId, state: AssetState<Balance>) -> DispatchResult {
+		ensure!(!Assets::<T>::contains_key(asset_id), Error::<T>::AssetAlreadyAdded);
+		ensure!(T::AssetRegistry::exists(asset_id), Error::<T>::AssetNotRegistered);
+
+		<Assets<T>>::insert(asset_id, state);
+
+		Ok(())
+	}
+
+	/// Updates states of 2 non-hub assets given calculated trade result.
+	#[require_transactional]
+	pub fn update_omnipool_state_given_trade_result(
+		asset_in: T::AssetId,
+		asset_out: T::AssetId,
+		trade: TradeStateChange<Balance>,
+	) -> DispatchResult {
+		let delta_hub_asset = trade
+			.asset_in
+			.delta_hub_reserve
+			.merge(
+				trade
+					.asset_out
+					.delta_hub_reserve
+					.merge(BalanceUpdate::Increase(trade.hdx_hub_amount))
+					.ok_or(ArithmeticError::Overflow)?,
+			)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		match delta_hub_asset {
+			BalanceUpdate::Increase(val) if val == Balance::zero() => {
+				// nothing to do if zero.
+			}
+			BalanceUpdate::Increase(_) => {
+				// trade can only burn some.
+				return Err(Error::<T>::HubAssetUpdateError.into());
+			}
+			BalanceUpdate::Decrease(amount) => {
+				T::Currency::withdraw(T::HubAssetId::get(), &Self::protocol_account(), amount)?;
+			}
+		};
+
+		Self::update_imbalance(trade.delta_imbalance)?;
+
+		Self::update_asset_state(asset_in, trade.asset_in)?;
+		Self::update_asset_state(asset_out, trade.asset_out)?;
+
+		Self::update_hdx_subpool_hub_asset(trade.hdx_hub_amount)?;
+
+		Ok(())
+	}
+
+	/// Updates states of anasset given calculated trade result where HUb asset was traded.
+	#[require_transactional]
+	pub fn update_omnipool_state_given_hub_asset_trade(
+		asset: T::AssetId,
+		trade: HubTradeStateChange<Balance>,
+	) -> DispatchResult {
+		Self::update_imbalance(trade.delta_imbalance)?;
+		Self::update_asset_state(asset, trade.asset)?;
+		Ok(())
+	}
+
+	/// Load state of an asset and update it with given delta changes.
+	pub fn update_asset_state(asset_id: T::AssetId, delta: AssetStateChange<Balance>) -> DispatchResult {
+		let state = Self::load_asset_state(asset_id)?;
+		let updated_state = state.delta_update(&delta).ok_or(ArithmeticError::Overflow)?;
+		Self::set_asset_state(asset_id, updated_state);
+
+		Ok(())
+	}
+
+	/// Load position and check its owner
+	/// Returns Forbidden if not position owner
+	pub fn load_position(
```

Comment:
I would rewrite it to use optional `maybe_check_owner` parameter. But I also don't have any problem to leave it as it is.

Reply:
can be refactor later when it would be need it, i dont see the need atm